File: clients/go/options.go
package queuebox

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

// Logger is the logger the library writes to. It prints nothing without one.
type Logger interface {
	Info(message string, fields map[string]any)
	Warn(message string, fields map[string]any)
	Error(message string, fields map[string]any)
}

type noopLogger struct{}

func (noopLogger) Info(string, map[string]any)  {}
func (noopLogger) Warn(string, map[string]any)  {}
func (noopLogger) Error(string, map[string]any) {}

// Options holds the settings of one worker.
type Options struct {
	// Source is the source whose messages this worker takes. It is mandatory.
	Source string
	// BatchSize is the largest number of messages one claim takes. The default is ten.
	BatchSize int
	// LeaseMS is the lease duration in milliseconds. The renewal runs every third of it.
	// The default is thirty thousand.
	LeaseMS int
	// MaxConcurrency is the largest number of handlers that run at one time. It never passes the
	// batch size. The default is the batch size.
	MaxConcurrency int
	// PollInterval is how long the worker waits after a claim that returned nothing.
	// The default is one second.
	PollInterval time.Duration
	// ShutdownGrace is how long a stop waits for the handlers that already run.
	// The default is thirty seconds.
	ShutdownGrace time.Duration
	// Dialect is the database dialect. The default is PostgreSQL.
	Dialect Dialect
	// Schema names the table and the columns. The default is the QueueBox V6 names.
	Schema *Schema
	// RetryPolicy decides what happens to a message whose handler returned an error.
	RetryPolicy RetryPolicy
	// Logger is the caller's logger.
	Logger Logger
}

type resolvedOptions struct {
	source          string
	batchSize       int
	leaseMS         int
	concurrency     int
	pollInterval    time.Duration
	shutdownGrace   time.Duration
	renewalInterval time.Duration
}

func (o Options) resolve() (resolvedOptions, Schema, RetryPolicy, Logger, error) {
	if strings.TrimSpace(o.Source) == "" {
		return resolvedOptions{}, Schema{}, nil, nil, errors.New("the source is mandatory")
	}

	batchSize := o.BatchSize
	if batchSize == 0 {
		batchSize = 10
	}

	leaseMS := o.LeaseMS
	if leaseMS == 0 {
		leaseMS = 30_000
	}

	pollInterval := o.PollInterval
	if pollInterval == 0 {
		pollInterval = time.Second
	}

	shutdownGrace := o.ShutdownGrace
	if shutdownGrace == 0 {
		shutdownGrace = 30 * time.Second
	}

	if batchSize < 0 {
		return resolvedOptions{}, Schema{}, nil, nil, fmt.Errorf("the batch size %d is not positive", batchSize)
	}

	// The renewal runs every third of the lease, so a shorter lease has no interval at all.
	if leaseMS < 3 {
		return resolvedOptions{}, Schema{}, nil, nil, fmt.Errorf("the lease of %d milliseconds leaves no room for a renewal", leaseMS)
	}

	if o.MaxConcurrency < 0 {
		return resolvedOptions{}, Schema{}, nil, nil, fmt.Errorf("the concurrency %d is not positive", o.MaxConcurrency)
	}

	if pollInterval < 0 {
		return resolvedOptions{}, Schema{}, nil, nil, errors.New("the poll interval is negative")
	}

	if shutdownGrace < 0 {
		return resolvedOptions{}, Schema{}, nil, nil, errors.New("the shutdown grace is negative")
	}

	concurrency := o.MaxConcurrency
	if concurrency == 0 {
		concurrency = batchSize
	}

	// Bounded memory: never hold more than the configured batch.
	concurrency = min(concurrency, batchSize)

	schema := DefaultSchema()
	if o.Schema != nil {
		schema = *o.Schema
	}

	policy := o.RetryPolicy
	if policy == nil {
		policy = NewDefaultRetryPolicy(5, time.Second, 5*time.Minute, 0.2)
	}

	var logger Logger = noopLogger{}
	if o.Logger != nil {
		logger = o.Logger
	}

	return resolvedOptions{
		source:          o.Source,
		batchSize:       batchSize,
		leaseMS:         leaseMS,
		concurrency:     concurrency,
		pollInterval:    pollInterval,
		shutdownGrace:   shutdownGrace,
		renewalInterval: max(time.Millisecond, time.Duration(leaseMS/3)*time.Millisecond),
	}, schema, policy, logger, nil
}

File: clients/typescript/src/options.ts
import type { InboxRetryPolicy } from "./retry.ts";
import type { InboxSchema, SqlDialect } from "./sql.ts";

/** The logger the library writes to. It prints nothing without one. */
export interface InboxLogger {
  info(message: string, fields?: Record<string, unknown>): void;
  warn(message: string, fields?: Record<string, unknown>): void;
  error(message: string, fields?: Record<string, unknown>): void;
}

/** The settings of one worker. */
export interface InboxOptions {
  /** The source whose messages this worker takes. It is mandatory. */
  readonly source: string;
  /** The largest number of messages one claim takes. */
  readonly batchSize?: number;
  /** The lease duration in milliseconds. The renewal runs every third of it. */
  readonly leaseMs?: number;
  /**
   * The largest number of handlers that run at one time. It never passes the batch size.
   * The default is one. A handler meets no sibling message unless the caller asks for
   * parallelism.
   */
  readonly maxConcurrency?: number;
  /** How long the worker waits after a claim that returned nothing. */
  readonly pollIntervalMs?: number;
  /** How long a shutdown waits for the handlers that already run. */
  readonly shutdownGraceMs?: number;
  /** The database dialect. */
  readonly dialect?: SqlDialect;
  /** The table and column names. */
  readonly schema?: InboxSchema;
  /** What happens to a message whose handler threw. */
  readonly retryPolicy?: InboxRetryPolicy;
  /** The caller's logger. */
  readonly logger?: InboxLogger;
}

/** The settings after the defaults are applied. */
export interface ResolvedOptions {
  readonly source: string;
  readonly batchSize: number;
  readonly leaseMs: number;
  readonly concurrency: number;
  readonly pollIntervalMs: number;
  readonly shutdownGraceMs: number;
  readonly dialect: SqlDialect;
  readonly renewalIntervalMs: number;
}

/**
 * Apply the defaults and reject a setting that cannot work.
 *
 * @param options the caller's settings
 * @returns the settings the worker applies
 */
export function resolveOptions(options: InboxOptions): ResolvedOptions {
  if (typeof options.source !== "string" || options.source.trim() === "") {
    throw new TypeError("The source is mandatory.");
  }

  const batchSize = options.batchSize ?? 10;
  const leaseMs = options.leaseMs ?? 30_000;
  const pollIntervalMs = options.pollIntervalMs ?? 1000;
  const shutdownGraceMs = options.shutdownGraceMs ?? 30_000;

  if (!Number.isInteger(batchSize) || batchSize <= 0) {
    throw new RangeError("The batchSize must be a positive whole number.");
  }

  // The renewal runs every third of the lease, so a shorter lease has no interval at all.
  if (!Number.isInteger(leaseMs) || leaseMs < 3) {
    throw new RangeError("The leaseMs must be a whole number of three or more.");
  }

  if (options.maxConcurrency !== undefined && (!Number.isInteger(options.maxConcurrency) || options.maxConcurrency <= 0)) {
    throw new RangeError("The maxConcurrency must be a positive whole number.");
  }

  if (pollIntervalMs <= 0) {
    throw new RangeError("The pollIntervalMs must be positive.");
  }

  if (shutdownGraceMs < 0) {
    throw new RangeError("The shutdownGraceMs must not be negative.");
  }

  return {
    source: options.source,
    batchSize,
    leaseMs,
    // Bounded memory: never hold more than the configured batch.
    concurrency: Math.min(options.maxConcurrency ?? 1, batchSize),
    pollIntervalMs,
    shutdownGraceMs,
    dialect: options.dialect ?? "postgresql",
    renewalIntervalMs: Math.max(1, Math.floor(leaseMs / 3)),
  };
}
